libevmasm/Assembly.cpp

```diff
@@ -53,6 +53,54 @@ using namespace solidity::evmasm;
 using namespace solidity::langutil;
 using namespace solidity::util;
 
+namespace
+{
+
+/// Produces instruction location info in RAII style. When an assembly instruction is added to the bytecode,
+/// this class can be instantiated in that scope. It will record the current bytecode size (before addition)
+/// and, at destruction time, record the new bytecode size. This information is then added to an external
+/// instruction locations vector.
+/// If the instruction decomposes into multiple individual evm instructions, `emit` can be
+/// called for all but the last one (which will be emitted by the destructor).
+class InstructionLocationEmitter
+{
+public:
+	InstructionLocationEmitter(
+		std::vector<LinkerObject::InstructionLocation>& _instructionLocations,
+		bytes const& _bytecode,
+		size_t const _assemblyItemIndex
+	):
+		m_instructionLocations(_instructionLocations),
+		m_bytecode(_bytecode),
+		m_assemblyItemIndex(_assemblyItemIndex),
+		m_instructionLocationStart(_bytecode.size())
+	{}
+
+	~InstructionLocationEmitter()
+	{
+		emit();
+	}
+
+	void emit()
+	{
+		auto const end = m_bytecode.size();
+		m_instructionLocations.emplace_back(LinkerObject::InstructionLocation{
+			.start = m_instructionLocationStart,
+			.end = end,
+			.assemblyItemIndex = m_assemblyItemIndex
+		});
```

Comment:
Shouldn't this be as follows:
```suggestion
		m_instructionLocations.emplace_back(
			m_instructionLocationStart,
			end,
			m_assemblyItemIndex
		);
```
Otherwise it's equivalent to a `push_back`, i.e. copied instead of constructed in place.

Reply:
yup but clang doesn't like that :(

Reply:
I'd suggest using `push_back()` in that situation though. The whole shtick of `emplace_back()` is that it will perform an implicit conversion. When you know the exact type, you don't really want that.

Reply:
Generally I'd rather say the whole shtick is that stuff is created in-place (if possible) and we avoid an additional move. This is still the case if I know the precise type.
In our specific case there is exactly zero performance (or copy/move semantic) difference between `push_back` and `emplace_back` - both perform a move into an allocated memory region and that is it. For the sake of clarity I have changed it to `push_back`, though.

Reply:
But the creation in place happens via an implicit conversion (basically, though choosing a constructor, though it could be one with multiple args; which is why sometimes use `explicit` to prevent such conversions).

I'm just pointing this out because for some time I've seen a push to use the `emplace`-style methods everywhere, but `push_back()` actually seems to me like a better default, because it's stricter about what it accepts. I'd only use `emplace_back()` when I know I want such a conversion or construction in-place.

Reply:
I don't want to derail this PR and I think the push-back is great! Although I do have to push back a bit on the push-back. You absolutely can have an explicit constructor and still use `emplace_back` by virtue of its args. It is simply forwarded into `std::construct_at`. So this _can_ perform implicit conversion but doesn't have to :)